packages/schema/src/plugins/enhancer/enhance/index.ts

```diff
@@ -493,10 +504,72 @@ export function enhance(prisma: any, context?: EnhancementContext<${authTypePara
         // fix delegate payload union type
         source = this.fixDelegatePayloadType(typeAlias, delegateInfo, source);
 
+        // fix json field type
+        source = this.fixJsonFieldType(typeAlias, source);
+
         structure.type = source;
         return structure;
     }
 
+    private fixJsonFieldType(typeAlias: TypeAliasDeclaration, source: string) {
+        const modelsWithTypeField = this.model.declarations.filter(
+            (d): d is DataModel => isDataModel(d) && d.fields.some((f) => isTypeDef(f.type.reference?.ref))
+        );
+        const typeName = typeAlias.getName();
+
+        const getTypedJsonFields = (model: DataModel) => {
+            return model.fields.filter((f) => isTypeDef(f.type.reference?.ref));
+        };
+
+        const replacePrismaJson = (source: string, field: DataModelField) => {
+            return source.replace(
+                new RegExp(`(${field.name}\\??\\s*):[^\\n]+`),
+                `$1: ${field.type.reference!.$refText}${field.type.array ? '[]' : ''}${
+                    field.type.optional ? ' | null' : ''
+                }`
+            );
+        };
+
+        // fix "$[Model]Payload" type
+        const payloadModelMatch = modelsWithTypeField.find((m) => `$${m.name}Payload` === typeName);
```

Comment:
Instead of replacement, could we slightly alter the type construction, to support type augmentation?

https://github.com/prisma/prisma/issues/3219#issuecomment-2400772675

Reply:
Hi @nwidynski , I'm glad you found the work here 😄.

I think your proposal is a good approach to making the model types easily customizable. Can it be simplified to the following (without the nested `$user`)?

```ts
export interface $UserExtensionScalar {}

export interface $UserScalar {
    // original scalar fields
    ...
}

export type $UserPayload<ExtArgs extends $Extensions.InternalArgs = $Extensions.DefaultArgs> = {
    ...
    scalars: $Extensions.GetPayloadResult<
      Omit<$UserScalar, keyof $UserExtensionScalar> & $UserExtensionScalar,
      ExtArgs["result"]["user"]>
}


// user land
import '@prisma/client';

declare module '@prisma/client' {
    namespace Prisma {
        export interface $UserExtensionScalar {
            metadata: { ... }
        }
    }
}
```

Curious about your use case: when ZenStack's native support for typed-json-field is available, do you still need this augmentation-based extension?

Reply:
Hey,

unfortunately the nested property is necessary as typescript can’t override existing declarations but only patch them. In order to make conditionals on those patches, a known property must be patched. It can have any name though, it doesnt have to be $user.

As for your second question, yes I would still have a use case for them in order to get strongly typed model results while they are separated in different extensions. This currently doesnt work well, especially for nested includes.

Reply:
Got it. For the field-type patching, I was thinking of "omit then patch": `Omit<$UserScalar, keyof $UserExtensionScalar> & $UserExtensionScalar`. I haven't tried it yet though.

My current goal is to make the "type" declaration a preview feature in v2.8.0 to get more feedback about the direction. I think we can incorporate your suggestion after the general approach is confirmed viable.

Reply:
Yes, that delivers additional convenience, although the $UserScalar could also be reused in user land to build the extension.